### ISSUE TITLE
Super-linter algorithm for finding changed files is broken

### ISSUE THREAD

**Issue by dec5e** [super-linter/super-linter#1167]
## Describe the bug

`git diff-tree` used by linter on `push` checks only files from the last commit and regardless of files status, so deleted or renamed files are also checked.  

There are now 2 different `git` commands used for finding the list of broken files (find them here: https://github.com/github/super-linter/blob/v3.14.4/lib/functions/buildFileList.sh#L59-L105). `git diff-tree` is used on `push`, `git diff` is used otherwise. A separate behaviour for push was added in #905 and follow-up #1049 to fix push to a default branch. But it seems working only in rare special cases: only one commit is pushed and no files are deleted. Otherwise wrong list is reported.

## Expected behavior

All changed non-deleted files from all commits added on top of the default branch should be checked, deleted files should be ignored.

## Steps to Reproduce

Here are the steps to reproduce within git (super-linter is not involved.

Adding first file, everything works fine:
```bash
+ git log --oneline master...465e71b
465e71b (HEAD -> test) Add test1.txt

+ git diff-tree --no-commit-id --name-only -r 465e71b
test/test1.txt

+ git diff --name-only master...465e71b --diff-filter=d
test/test1.txt
```

Now adding the second file in a separate commit, `diff git-tree` reports only the last file - only files changed in the last commit:
```bash
+ git log --oneline master...1c3d652
1c3d652 (HEAD -> test) Add test2.txt
465e71b Add test1.txt

+ git diff-tree --no-commit-id --name-only -r 1c3d652
test/test2.txt

+ git diff --name-only master...1c3d652 --diff-filter=d
test/test1.txt
test/test2.txt
```

Renaming first file, `git diff-tree` reports again only files from the last commit and as another issue - reports deleted file as well:
```bash
+ git log --oneline master...f5617a2
f5617a2 (HEAD -> test) Rename test1.txt to test3.txt
1c3d652 Add test2.txt
465e71b Add test1.txt

+ git diff-tree --no-commit-id --name-only -r f5617a2
test/test1.txt
test/test3.txt

+ git diff --name-only master...f5617a2 --diff-filter=d
test/test2.txt
test/test3.txt
```

Now deleting the second file, `git diff-tree` reports only this deleted file:
```shell
+ git log --oneline master...ee90eaf
ee90eaf (HEAD -> test) Remove test2.txt
f5617a2 Rename test1.txt to test3.txt
1c3d652 Add test2.txt
465e71b Add test1.txt

+ git diff-tree --no-commit-id --name-only -r ee90eaf
test/test2.txt

+ git diff --name-only master...ee90eaf --diff-filter=d
test/test3.txt
```

## Additional context

Right now `git diff` command looks better for all cases except finding changes for a push to a `DEFAULT_BRANCH` (the case described in #900). I'd say `git diff` should be used instead of the other command, or `git diff-tree` needs to be called with a right list of parameters. 

I don't have suggestions on how better to fix the original issue from #900. If there needs to be a special code condition for that case maybe it should be checking not only git `push` event but also if current branch is a `DEFAULT_BRANCH`. 

This project really needs testing for the "get diff files list" behaviour. It was broken couple of times already. Having wrong files checked is really disappointing. What makes it even more disappointing is that you often don't even realize this. With current broken behaviour wrong files were linted in the last 3 months when `push` event was used. I believe that's quite a lot of people and projects since this setup is recommended by super-linter readme :cry: 

Additional tough thing is that it's impossible to debug "get diff files list" in local, since `VALIDATE_ALL_CODEBASE=true` behaviour is used in that case.

**Comment by dec5e** [comment]
I solved this for now in my projects with a switch to `pull_request` GitHub event. This event uses correct git command and a valid list of files for checks. But works obviously only for commits pushed to Pull Requests, doesn't work for commits pushed to a branch without PR opened:
```yaml
on:
  pull_request:
    branches: [master]
```

**Comment by dec5e** [comment]
Ping @admiralAwkbar @GaboFDC @ferrarimarco @zkoppert. Could you check please, I think that's a very important issue.

**Comment by zkoppert** [comment]
Yes it looks like using a `git diff --diff-filter` may be appropriate here to essentially ignore deleted files by only including all other types (modified, new, etc.).

**Comment by zkoppert** [comment]
Anybody got a moment to open a draft pull request and start playing around with this? If not, I may have a little time later today.

**Comment by admiralAwkbar** [comment]
@zkoppert @dec5e i think we just implemented new find features in the latest branches
can you validate?

**Comment by dec5e** [comment]
@admiralAwkbar in `master`? could point more directly where to look for changes?

**Comment by dec5e** [comment]
If you refer to #1259 - I don't think it has fixed the issue I described here. As I wrote `git diff-tree $SHA_COMMIT` has only limited usage for the case of finding a list of changed files, namely, when we need changes from the exact commit. That might work for single commit pushed to main or any other branch but wouldn't work for multi commit changes, at least without changes to command parameters.

`git diff master...GITHUB_SHA` fetches files from many commits, `git diff-tree $GITHUB_SHA` fetches only changes made in `GITHUB_SHA` commit.

**Comment by GaboFDC** [comment]
So if i'm getting this correctly, the problem is when you push multiple commits, and you are not opening a PR?
Could you detail a bit more about this workflow? so we understand the use case better?

Ans you stated we use git diff when the trigger is a PR event, so as you stated, using the correct trigger is the solution.

In the readme we use both triggers, we don't recommend an specific one.

**Comment by dec5e** [comment]
My ideal workflow is: create a local branch, make several commits, push to a remote, check linter results, fix issues and push more commits with fixes if needed, only then create a pull request. To make this work I need to use `push` trigger. Right now I can't do so and I'm forced to use `pull_request` trigger.

> In the readme we use both triggers, we don't recommend an specific one.
Do you mean I have to use both?

I don't want to run checks twice and also I liked `push` over `pull_request` so every time I push to a branch a check is started and I don't need to create a PR immediately, I can create it once I'm ready with the changes and when I fixed all linter complains.

**Comment by dec5e** [comment]
The problem is independent from do I create a PR or not, the problem is with `push` trigger itself. The git command `diff-tree` used for `push` with the current list of parameters works only for a single commit, plus it doesn't exclude deleted files. Adding more workarounds and trying to catch error is an invalid way imho, proper parameters need to be passed so all the checks "file is in git metadata but doesn't exist" will become unnecessary. And as said, these workarounds doesn't solve issues with multi commit pushes.

**Comment by dec5e** [comment]
So the problem I'm trying to denote here: the git command used for `push` trigger does not cover all possible cases and needs to be improved. I see 2 possible ways for the improvement: take more than 1 commit into account and ignore deleted files (like it's done in another git command used in `pull_request` trigger with `--diff-filter=d`). 

I don't have time unfortunately to properly test and suggest a solution myself.

**Comment by pap0135** [comment]
![image](https://user-images.githubusercontent.com/72086601/113497089-f35a7200-9508-11eb-9e0f-0ab3058b4e27.jpeg)

**Comment by pap0135** [comment]
****

**Comment by dec5e** [comment]
This is not solved yet I guess

**Comment by lex2491** [comment]
Like

**Comment by lex2491** [comment]
@lex2491

**Comment by sobigyan** [comment]
I believe this issue is still persistent

**Comment by l4kr** [comment]
So it's not a me issue, got it.

**Comment by ferrarimarco** [comment]
Hi folks! This issue has been open for quite some time.

Do we have clear reproduction cases when using the latest super-linter release?

**Comment by ferrarimarco** [comment]
I've added a few tests and a possible fix in #5044.

Thanks @dec5e for the reproduction case.
